source/characterProcessing.py

```diff
@@ -540,15 +540,34 @@ def _regexpRepl(self, m):
 			if group == "simple":
 				# Simple symbol.
 				symbol = self.computedSymbols[text]
+				replacement = symbol.replacement
 			else:
 				# Complex symbol.
 				index = int(group[1:])
 				symbol = self._computedComplexSymbolsList[index]
+				replacement = ''
+
+				escape = False
+				for c in symbol.replacement:
+					if not escape:
+						if c == '\\':
+							escape = True
+						else:
+							replacement += c
+					else:
+						if c == '\\':
+							replacement += '\\'
+						elif c >= '0' and c <= '9':
+							replacement += m.group(m.lastindex + ord(c) - ord('0'))
+						else:
+							log.error("Invalid reference \\%s" % c)
+						escape = False

```

Comment:
Please extract this block to a named function. Did you look to see if the regex module can handle these escapes directly? I'm concerned this approach will be error prone. If this must be done manually, please add unit tests.

Reply:
Ah, let me check, it seems like m.expand could do it.

Reply:
No, it doesn't work, because the numbers of the references that are used by `m.expand` are relative to the whole regexp, and not relative to the regexp inside the named group.

Reply:
(i.e. it's the `m+lastindex + ` here which is missing in the `re`-provided implementation)

Reply:
Is the dic engine actually tested at all? I don't find anything existing in the repository.

Reply:
I have added a test, but I don't know why the python linter doesn't like the heading comments, which are just exactly like in other files.

(there is also a fix in the french locale file, I have contacted the translator to get it fixed ahead of this)

Reply:
Commends in other files were mostly introduced before we introduced the linter unfortunately. You need to insert space after each hash sign to make it happy and add one more blank line before 
`class TestComplex(unittest.TestCase):`

Also were no longer placing the file name as the first line in the copyright header so please remove it.

Reply:
Ah, the linter only looks at new code, that's why. Ok, pushed a fixed version, thanks for the explanation.